src/prefect/agent/kubernetes/agent.py

```diff
@@ -93,9 +95,139 @@ def __init__(
             config.load_kube_config()
 
         self.batch_client = client.BatchV1Api()
+        self.core_client = client.CoreV1Api()
+        self.k8s_client = client
+
+        self.jobs = self._get_current_prefect_jobs()
 
         self.logger.debug(f"Namespace: {self.namespace}")
 
+    def _get_current_prefect_jobs(self) -> list:
+        """
+        Get a list of prefect jobs that are currently present in the cluster.
+
+        Returns:
+            - A list of dictionaries where each entry matches a job name, its pod names, and flow run id
+        """
+
+        jobs_list = list()
+
+        self.logger.debug(
+            "Retrieving information of jobs that are currently in the cluster..."
+        )
+
+        more = True
+        _continue = ""
+        while more:
+            try:
+                jobs = self.batch_client.list_namespaced_job(
+                    namespace=self.namespace,
+                    label_selector="prefect.io/identifier",
+                    limit=20,
+                    _continue=_continue,
+                )
+                _continue = jobs.metadata._continue
+                more = bool(_continue)
+
+                for job in jobs.items:
+                    pods = self.core_client.list_namespaced_pod(
+                        namespace=self.namespace,
+                        label_selector="prefect.io/identifier={}".format(
+                            job.metadata.labels.get("prefect.io/identifier")
+                        ),
+                    )
+
+                    pod_names = []
+                    for pod in pods.items:
+                        pod_names.append(pod.metadata.name)
+
+                    jobs_list.append(
+                        {
+                            "job_name": job.metadata.name,
+                            "pod_names": pod_names,
+                            "flow_run_id": job.metadata.labels.get(
+                                "prefect.io/flow_run_id"
+                            ),
+                        }
+                    )
+            except self.k8s_client.rest.ApiException as exc:
+                if exc.status == 410:
+                    self.logger.debug("List jobs continue token expired, relisting")
+                    _continue = ""
+                    continue
+
+        self.logger.debug(
+            f"Found {len(jobs_list)} existing prefect jobs in the cluster"
+        )
+        return jobs_list
+
+    def heartbeat(self) -> None:
```

Comment:
This loop makes quite a few calls to the k8s api, and only does `get` operations based on existing job/pod records stored on the agent. This has a few consequences:

- Additional pods started by a job post submit (race condition noted below) won't be caught or checked
- I don't believe this handles the job-starts-a-job (as done currently with the `KubernetesJobEnvironment`) since the subjob won't be stored in the `jobs` mapping.

A different approach would be to drop all local state and either:
- do a periodic full relist of all jobs in the namespace (with appropriate label selectors)
- watch the namespace for job/pod events (again, with appropriate label selectors). This is common for managing infrastructure (it's what we do for dask-gateway and jupyterhub), which would reduce api calls to the k8s api server at the cost of increased local complexity. If `heartbeat` runs on the minute -> minutes scale this complexity likely isn't needed yet.

Pseudocode with a loop:

```python
def heartbeat(self):
    for job in self.generator_of_jobs_in_namespace_matching_label_selector():
        if self.is_done(job) or self.has_pods_stuck_in_bad_state(job):
            self.delete_job(job)
```

Either way, I think we should try to avoid holding onto local state for what jobs/pods to manage and rely on k8s api patterns to handle this instead.

Reply:
Interesting, so the heartbeat here won't run on the minute -> minutes scale and instead looks to run on the same iteration cycle as the normal querying for flow runs (10s). I haven't used watch events in the k8s python client before so I'll have to look into how to perform it